reactos/drivers/usb/usbohci_new/usbohci.c

```diff
@@ -0,0 +1,2752 @@
+                    IN PVOID ohciEndpoint,
+                    IN PVOID transferParameters,
+                    IN PVOID ohciTransfer,
+                    IN PVOID sgList)
+{
+    POHCI_EXTENSION OhciExtension = ohciExtension;
+    POHCI_ENDPOINT OhciEndpoint = ohciEndpoint;
+    PUSBPORT_TRANSFER_PARAMETERS TransferParameters = transferParameters;
+    POHCI_TRANSFER OhciTransfer = ohciTransfer;
+    PUSBPORT_SCATTER_GATHER_LIST SGList = sgList;
+    ULONG TransferType;
+
+    DPRINT_OHCI("OHCI_SubmitTransfer: ... \n");
+
+    RtlZeroMemory(OhciTransfer, sizeof(OHCI_TRANSFER));
+
+    OhciTransfer->TransferParameters = TransferParameters;
+    OhciTransfer->OhciEndpoint = OhciEndpoint;
+
+    TransferType = OhciEndpoint->EndpointProperties.TransferType;
+
+    if (TransferType == USBPORT_TRANSFER_TYPE_CONTROL)
+    {
+        return OHCI_ControlTransfer(OhciExtension,
+                                    OhciEndpoint,
+                                    TransferParameters,
+                                    OhciTransfer,
+                                    SGList);
+    }
+
+    if (TransferType == USBPORT_TRANSFER_TYPE_BULK ||
+        TransferType == USBPORT_TRANSFER_TYPE_INTERRUPT)
+    {
+        return OHCI_BulkOrInterruptTransfer(OhciExtension,
+                                            OhciEndpoint,
+                                            TransferParameters,
+                                            OhciTransfer,
+                                            SGList);
+    }
+
+    return MP_STATUS_FAILURE;
+}
+
+MPSTATUS
+NTAPI
+OHCI_SubmitIsoTransfer(IN PVOID ohciExtension,
+                       IN PVOID ohciEndpoint,
+                       IN PVOID transferParameters,
+                       IN PVOID ohciTransfer,
+                       IN PVOID isoParameters)
+{
+    DPRINT1("OHCI_SubmitIsoTransfer: UNIMPLEMENTED. FIXME\n");
+    return MP_STATUS_SUCCESS;
+}
+
+VOID
+NTAPI
+OHCI_ProcessDoneTD(IN POHCI_EXTENSION OhciExtension,
+                   IN POHCI_HCD_TD TD,
+                   IN BOOLEAN IsPortComplete)
+{
+    POHCI_TRANSFER OhciTransfer;
+    POHCI_ENDPOINT OhciEndpoint;
+    ULONG_PTR Buffer;
+    ULONG_PTR BufferEnd;
+    ULONG Length;
+
+    DPRINT_OHCI("OHCI_ProcessDoneTD: ... \n");
+
+    OhciTransfer = TD->OhciTransfer;
+    OhciEndpoint = OhciTransfer->OhciEndpoint;
+
+    OhciTransfer->PendingTDs--;
+
+    Buffer = (ULONG_PTR)TD->HwTD.gTD.CurrentBuffer;
+    BufferEnd = (ULONG_PTR)TD->HwTD.gTD.BufferEnd;
+
+    if (TD->Flags & OHCI_HCD_TD_FLAG_NOT_ACCESSED)
+    {
+        TD->HwTD.gTD.Control.ConditionCode = OHCI_TD_CONDITION_NO_ERROR;
+    }
+    else
+    {
+        if (TD->HwTD.gTD.CurrentBuffer)
+        {
+            if (TD->TransferLen)
+            {
+                Length = (BufferEnd & (PAGE_SIZE - 1)) -
+                         (Buffer & (PAGE_SIZE - 1));
+
+                Length++;
+
+                if ((Buffer ^ BufferEnd) & ~(PAGE_SIZE - 1))
```

Comment:
This is pretty obscure. `Buffer >> PAGE_SHIFT != BufferEnd >> PAGE_SHIFT` should achieve the same thing but seems much more readable.

Reply:
Changed.